k2/csrc/fsa_utils.cu

```diff
@@ -177,101 +343,145 @@ static Fsa AcceptorFromStream(std::string first_line, std::istringstream &is,
   return fsa;
 }
 
-static Fsa TransducerFromStream(std::string first_line, std::istringstream &is,
-                                bool openfst, Array1<int32_t> *aux_labels) {
+/* Create a transducer from a stream, assuming the transducer is in the OpenFST
+   format:
+
+   src_state1 dest_state1 label1 aux_label1 score1
+   src_state2 dest_state2 label2 aux_label2 score2
+   ... ...
+   final_state final_score
+
+   We will negate the cost/score when we read them in. Also note, OpenFST may
+   omit the cost/score if it is 0.0.
+
+   If there are multiple final states, we will create a super final state, and
```

Comment:
I don't believe it is actually conditional.
The code as you have it requires there to be at least one final state, no?
I think it would be better to always create a superfinal state, but not require there to be any final states.
(would just be disconnected in that case).

Reply:
I thought about it for a second, but felt an FSA without a final state will be use less?

Let's do this then. We will always add a super final state. If there's at least one final state in the original FSA, we create an arc(s) from the original final state(s) to the super final state, with label -1; if there's no final state in the original FSA, we add the super final state (max_id + 1) but with no arcs connecting it with the original FSA. Or do you think we should add an arc from max_id to the super final state with label -1 and weight 0?

Reply:
Always add super-final state but no special cases, we don't want to guess at what the user wants.  Better to have predictable behavior.  Non-coaccessible FSAs do sometimes arise naturally.

Reply:
I removed the constrain of the presence of the final state in the OpenFST format. Super final state will always be added.